atef/widgets/ophyd.py

```diff
@@ -0,0 +1,729 @@
+
+    def _update_attr(self, attr: str):
+        """Update an attribute of the device."""
+        setpoint = None
+        data = self.data[attr]
+
+        try:
+            if not data.signal.connected:
+                return
+        except TimeoutError:
+            return
+
+        if not data.description:
+            try:
+                data.description = data.signal.describe()[data.signal.name] or {}
+            except Exception:
+                data.description = {
+                    "units": data.signal.metadata.get("units", ""),
+                }
+
+        try:
+            get_setpoint = getattr(data.signal, "get_setpoint", None)
+            if callable(get_setpoint):
+                setpoint = get_setpoint()
+            else:
+                setpoint = getattr(data.signal, "setpoint", None)
+            readback = data.signal.get()
+        except TimeoutError as ex:
+            # Don't spam on failure to connect
+            logger.debug("Failed to connect to %s.%s (%s)", self.device.name, attr, ex)
+            return
+        except Exception:
+            logger.exception(
+                "Poll thread for %s.%s @ %.3f sec failure",
+                self.device.name,
+                attr,
+                self.poll_rate,
+            )
+            self._attrs.remove(attr)
+            return
+
+        new_data: Dict[str, Any] = {}
+        if readback is not None:
+            units = data.description.get("units", "") or ""
+            new_data["readback"] = f"{readback} {units}"
+        if setpoint is not None:
+            new_data["setpoint"] = setpoint
+
+        for key, value in new_data.items():
+            old_value = getattr(data, key)
+
+            try:
+                changed = np.any(old_value != value)
+            except Exception:
+                ...
+            else:
+                if changed or old_value is None:
+                    for key, value in new_data.items():
+                        setattr(data, key, value)
+                    self.data_changed.emit(attr)
+                    return
+
+    def run(self):
+        """The thread polling loop."""
+        self.running = True
+        self._attrs = self._instantiate_device()
+
+        # We may have already created the dictionary; only do it if necessary:
+        if not self.data:
+            self.data.update(**OphydAttributeData.from_device(self.device))
+
+        self.data_ready.emit()
+
+        while self.running:
+            t0 = time.monotonic()
+            for attr in list(self._attrs):
+                self._update_attr(attr)
+                if not self.running:
+                    break
+                time.sleep(0)
+
+            if self.poll_rate <= 0.0:
+                # A zero or below means "single shot" updates.
+                break
+
+            elapsed = time.monotonic() - t0
+            time.sleep(max((0, self.poll_rate - elapsed)))
+
+
+class PolledDeviceModel(QtCore.QAbstractTableModel):
```

Comment:
Absolutely worth pulling this and the other ophyd Qt stuff out into a more central module when we're done trialing them here in atef. :+1: